invenio_sequencegenerator/models.py

```diff
@@ -0,0 +1,202 @@
+
+"""Models for storing counters for each pattern."""
+
+import re
+
+from invenio_db import db
+from pkg_resources import DistributionNotFound, get_distribution
+from sqlalchemy import ForeignKey
+from sqlalchemy.orm.exc import NoResultFound
+
+counter_regex = re.compile(r'({counter(!.)?(:.*)?})')
+
+
+class Sequence(db.Model):
+    """Represents sequence definitions."""
+
+    __tablename__ = 'sequencegenerator_sequence'
+
+    name = db.Column(db.String, primary_key=True)
+    """The identifying name of the sequence."""
+
+    template = db.Column(db.String, unique=True)
+    """The template of the sequence."""
+
+    parent = db.Column(ForeignKey(name))
+    """Indicates if the template depends on another one."""
+
+    @classmethod
+    def create(cls, name, template):
+        """Create a new sequence."""
+        placeholders = extract_placeholders(template)
+
+        # Check that {counter} exists and double its brackets
+        if all(map(
+                lambda p: not counter_regex.search(
+                    '{}{}{}'.format('{', p, '}')
+                ), placeholders
+        )):
+            raise InvalidTemplate('No counter placeholder')
+        template = double_counter(template)
+
+        # Check if parent sequence exists
+        names = [seq.name for seq in Sequence.query.all()]
+        parents = filter(lambda p: p in names, placeholders)
+        if len(parents) > 1:
+            raise InvalidTemplate('More than 1 parents in template')
+
+        parent = parents[0] if parents else None
+
+        db.session.add(cls(
+            name=name,
+            template=template,
+            parent=parent
+        ))
+
+    @classmethod
+    def next(cls, id, **kwargs):
+        """Get the next report number for the given sequence."""
+        with db.session.begin_nested():
+            seq = db.session.query(cls).filter_by(name=id).first() or \
+                  db.session.query(cls).filter_by(template=id).first()
+            if seq is None:
+                raise NoSuchSequence()
+
+            try:
+                template = seq.template.format(**kwargs)
+            except KeyError:
+                raise InvalidCall('Improper arguments for string formatting')
+
+            if seq.parent:
+                # Check if parent's report number exists in PID store
+                try:
+                    get_distribution('invenio-pidstore')
+                    # from invenio_pidstore.errors import PIDDoesNotExistError
+                    from invenio_pidstore.models import PersistentIdentifier
+
+                    try:
+                        PersistentIdentifier.query.filter_by(
+                            pid_value=template
+                        ).one()
+                    except NoResultFound:
+                        raise InvalidCall('Parent PID does not exist')
+                except DistributionNotFound:
+                    pass
+
+                return ReportNumber.next_or_create(template)
```

Comment:
The code in `if` and `else` branch ends with `return` which is exactly the same. Can you move it to top method level?

```
if cond:
    # ...
    return foo
else:
    return foo

# translates to 

if cond:
    # ...
return foo
```